File: WinForms/VB/EditorsMainDemo/Data/MoviesData.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.Internal

Namespace DevExpress.XtraEditors.Demos.Data

    Public Module MoviesData

        <System.FlagsAttribute>
        Public Enum MovieGenre
            None = 0
            Action = &H1
            Adventure = &H2
            Animation = &H4
            Biography = &H8
            Comedy = &H10
            Crime = &H20
            Documentary = &H40
            Drama = &H80
            Family = &H100
            Fantasy = &H200
            History = &H400
            Horror = &H800
            Music = &H1000
            Musical = &H2000
            Mystery = &H4000
            Romance = &H8000
            SciFi = &H10000
            Sport = &H20000
            Thriller = &H40000
            War = &H80000
            Western = &H100000
        End Enum

        Private XmlDataSet As System.Data.DataSet

        Public Producers As System.Collections.Generic.IList(Of DevExpress.XtraEditors.Demos.Data.MoviesData.Producer)

        Public Movies As System.Collections.Generic.IList(Of DevExpress.XtraEditors.Demos.Data.MoviesData.Movie)

        Public Sub Load()
            Call DevExpress.XtraEditors.Demos.Data.MoviesData.InitDataSet()
            If DevExpress.XtraEditors.Demos.Data.MoviesData.Producers Is Nothing Then DevExpress.XtraEditors.Demos.Data.MoviesData.Producers = DevExpress.XtraEditors.Demos.Data.MoviesData.LoadProducers().ToList()
            If DevExpress.XtraEditors.Demos.Data.MoviesData.Movies Is Nothing Then DevExpress.XtraEditors.Demos.Data.MoviesData.Movies = DevExpress.XtraEditors.Demos.Data.MoviesData.LoadMovies().ToList()
        End Sub

        Private Sub InitDataSet()
            If DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet Is Nothing Then
                DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet = New System.Data.DataSet()
                Call DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.ReadXml(DevExpress.Internal.DataDirectoryHelper.GetDataFile("VideoRent.xml"))
            End If
        End Sub

        Private Function LoadMovies() As IEnumerable(Of DevExpress.XtraEditors.Demos.Data.MoviesData.Movie)
            Dim tableMovie = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("Movie"), System.Data.DataTable)
            Dim tableMovieArtist = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("MovieArtist"), System.Data.DataTable)
            Dim tableMovieArtistLine = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("MovieArtistLine"), System.Data.DataTable)
            Dim tablePerson = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("Person"), System.Data.DataTable)
            Dim tableMovieCountry = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("MovieMovies_CountryCountries"), System.Data.DataTable)
            Dim tableCountry = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("Country"), System.Data.DataTable)
            Dim movie = From m In tableMovie.AsEnumerable() Join ma In tableMovieArtist.AsEnumerable() On m.Field(Of System.Guid)("Oid") Equals ma.Field(Of System.Guid)("Movie") Join mal In tableMovieArtistLine.AsEnumerable() On ma.Field(Of System.Guid)("Line") Equals mal.Field(Of System.Guid)("Oid") Join p In tablePerson.AsEnumerable() On ma.Field(Of System.Guid)("Artist") Equals p.Field(Of System.Guid)("Oid") Join mc In tableMovieCountry.AsEnumerable() On m.Field(Of System.Guid)("Oid") Equals mc.Field(Of System.Guid)("Movies") Join c In tableCountry.AsEnumerable() On mc.Field(Of System.Guid)("Countries") Equals c.Field(Of System.Guid)("Oid") Where System.Data.DataRowExtensions.Field(Of String)(mal, CStr(("Name"))).Equals("Director") Select New DevExpress.XtraEditors.Demos.Data.MoviesData.Movie() With {.MovieTitle = m.Field(Of String)("Title"), .Release = System.Data.DataRowExtensions.Field(Of System.DateTime)(m, CStr(("ReleaseDate"))).Year, .Genres = DevExpress.XtraEditors.Demos.Data.MoviesData.GetGenres(m.Field(Of Integer)("Genre")), .Producers = p.Field(Of String)("FirstName") & " " & p.Field(Of String)("LastName"), .Countries = c.Field(Of String)("Name")}
            Return From m In movie.AsEnumerable() Group m By __groupByKey1__ = New With {.MovieTitle = m.MovieTitle, .MovieRelease = m.Release, .MovieGenre = m.Genres} Into groupWithProducers = Group Select New DevExpress.XtraEditors.Demos.Data.MoviesData.Movie() With {.MovieTitle = __groupByKey1__.MovieTitle, .Release = __groupByKey1__.MovieRelease, .Genres = __groupByKey1__.MovieGenre, .Producers = String.Join(", ", groupWithProducers.[Select](Function(x) x.Producers).Distinct()), .Countries = String.Join(", ", groupWithProducers.[Select](Function(x) x.Countries).Distinct())}
        End Function

        Private Function LoadProducers() As IEnumerable(Of DevExpress.XtraEditors.Demos.Data.MoviesData.Producer)
            Dim tableMovieArtist = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("MovieArtist"), System.Data.DataTable)
            Dim tableMovieArtistLine = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("MovieArtistLine"), System.Data.DataTable)
            Dim tablePerson = TryCast(DevExpress.XtraEditors.Demos.Data.MoviesData.XmlDataSet.Tables("Person"), System.Data.DataTable)
            Return From ma In tableMovieArtist.AsEnumerable() Join mal In tableMovieArtistLine.AsEnumerable() On ma.Field(Of System.Guid)("Line") Equals mal.Field(Of System.Guid)("Oid") Join p In tablePerson.AsEnumerable() On ma.Field(Of System.Guid)("Artist") Equals p.Field(Of System.Guid)("Oid") Where System.Data.DataRowExtensions.Field(Of String)(mal, CStr(("Name"))).Equals("Director") Order By p.Field(Of String)("FirstName") & p.Field(Of String)("LastName") Select New DevExpress.XtraEditors.Demos.Data.MoviesData.Producer(p.Field(Of System.Guid)("Oid"), p.Field(Of String)("FirstName") & " " & p.Field(Of String)("LastName"))
        End Function

        Private Function GetGenres(ByVal value As Integer) As String
            Dim genres = CType(value, DevExpress.XtraEditors.Demos.Data.MoviesData.MovieGenre)
            Return genres.ToString("G")
        End Function

        Public Class Movie

            Public Property MovieTitle As String

            Public Property Release As Integer

            Public Property Genres As String

            Public Property Producers As String

            Public Property Countries As String
        End Class

        Public Class Producer

            Public Sub New(ByVal id As System.Guid, ByVal name As String)
                Me.Id = id
                Me.Name = name
            End Sub

            Public Property Id As Guid

            Public Property Name As String
        End Class
    End Module
End Namespace

File: WinForms/CS/EditorsMainDemo/Data/MoviesData.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Windows.Forms;
using DevExpress.Internal;

namespace DevExpress.XtraEditors.Demos.Data {
    public static class MoviesData {
        [Flags]
        public enum MovieGenre {
            None = 0, Action = 0x1, Adventure = 0x2, Animation = 0x4, Biography = 0x8, Comedy = 0x10, Crime = 0x20,
            Documentary = 0x40, Drama = 0x80, Family = 0x100, Fantasy = 0x200, History = 0x400, Horror = 0x800,
            Music = 0x1000, Musical = 0x2000, Mystery = 0x4000, Romance = 0x8000, SciFi = 0x10000, Sport = 0x20000,
            Thriller = 0x40000, War = 0x80000, Western = 0x100000
        }
        static DataSet XmlDataSet;
        public static IList<Producer> Producers;
        public static IList<Movie> Movies;
        public static void Load() {
            InitDataSet();
            if(Producers == null)
                Producers = LoadProducers().ToList();
            if(Movies == null)
                Movies = LoadMovies().ToList();
        }
        static void InitDataSet() {
            if(XmlDataSet == null) {
                XmlDataSet = new DataSet();
                XmlDataSet.ReadXml(DataDirectoryHelper.GetDataFile("VideoRent.xml"));
            }
        }
        static IEnumerable<Movie> LoadMovies() {
            var tableMovie = XmlDataSet.Tables["Movie"] as DataTable;
            var tableMovieArtist = XmlDataSet.Tables["MovieArtist"] as DataTable;
            var tableMovieArtistLine = XmlDataSet.Tables["MovieArtistLine"] as DataTable;
            var tablePerson = XmlDataSet.Tables["Person"] as DataTable;
            var tableMovieCountry = XmlDataSet.Tables["MovieMovies_CountryCountries"] as DataTable;
            var tableCountry = XmlDataSet.Tables["Country"] as DataTable;

            var movie = from m in tableMovie.AsEnumerable()
                        join ma in tableMovieArtist.AsEnumerable()
                        on m.Field<Guid>("Oid") equals ma.Field<Guid>("Movie")
                        join mal in tableMovieArtistLine.AsEnumerable()
                        on ma.Field<Guid>("Line") equals mal.Field<Guid>("Oid")
                        join p in tablePerson.AsEnumerable()
                        on ma.Field<Guid>("Artist") equals p.Field<Guid>("Oid")
                        join mc in tableMovieCountry.AsEnumerable()
                        on m.Field<Guid>("Oid") equals mc.Field<Guid>("Movies")
                        join c in tableCountry.AsEnumerable()
                        on mc.Field<Guid>("Countries") equals c.Field<Guid>("Oid")

                        where mal.Field<string>("Name").Equals("Director")
                        select new Movie() {
                            MovieTitle = m.Field<string>("Title"),
                            Release = m.Field<DateTime>("ReleaseDate").Year,
                            Genres = GetGenres(m.Field<int>("Genre")),
                            Producers = p.Field<string>("FirstName") + " " + p.Field<string>("LastName"),
                            Countries = c.Field<string>("Name")
                        };

            return from m in movie.AsEnumerable()
                   group m by new {
                       MovieTitle = m.MovieTitle,
                       MovieRelease = m.Release,
                       MovieGenre = m.Genres,
                   }
                      into groupWithProducers
                   select new Movie() {
                       MovieTitle = groupWithProducers.Key.MovieTitle,
                       Release = groupWithProducers.Key.MovieRelease,
                       Genres = groupWithProducers.Key.MovieGenre,
                       Producers = string.Join(", ", groupWithProducers.Select(x => x.Producers).Distinct()),
                       Countries = string.Join(", ", groupWithProducers.Select(x => x.Countries).Distinct()),
                   };
        }
        static IEnumerable<Producer> LoadProducers() {
            var tableMovieArtist = XmlDataSet.Tables["MovieArtist"] as DataTable;
            var tableMovieArtistLine = XmlDataSet.Tables["MovieArtistLine"] as DataTable;
            var tablePerson = XmlDataSet.Tables["Person"] as DataTable;
            return from ma in tableMovieArtist.AsEnumerable()
                   join mal in tableMovieArtistLine.AsEnumerable()
                   on ma.Field<Guid>("Line") equals mal.Field<Guid>("Oid")
                   join p in tablePerson.AsEnumerable()
                   on ma.Field<Guid>("Artist") equals p.Field<Guid>("Oid")
                   where mal.Field<string>("Name").Equals("Director")
                   orderby p.Field<string>("FirstName") + p.Field<string>("LastName")
                   select new Producer(p.Field<Guid>("Oid"), p.Field<string>("FirstName") + " " + p.Field<string>("LastName"));
        }
        static string GetGenres(int value) {
            var genres = (MovieGenre)value;
            return genres.ToString("G");
        }

        public class Movie {
            public string MovieTitle { get; set; }
            public int Release { get; set; }
            public string Genres { get; set; }
            public string Producers { get; set; }
            public string Countries { get; set; }
        }
        public class Producer {
            public Producer(Guid id, string name) {
                Id = id;
                Name = name;
            }
            public Guid Id { get; set; }
            public string Name { get; set; }
        }
    }
}
